certora/specs/Transfer.spec

```diff
@@ -0,0 +1,87 @@
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    function _.allowance(address, address) external => DISPATCHER(true);
+    function _.totalSupply() external => DISPATCHER(true);
+}
+
+ghost mapping(address => mathint) myBalances
+{
+    init_state axiom (forall address token. myBalances[token] == 0);
+}
+
+function summarySafeTransferFrom(address token, address from, address to, uint256 amount) {
+    if (from == currentContract) {
+        // Safe require because the reference implementation would revert.
+        myBalances[token] = require_uint256(myBalances[token] - amount);
+    }
+    if (to == currentContract) {
+        // Safe require because the reference implementation would revert.
+        myBalances[token] = require_uint256(myBalances[token] + amount);
+    }
+}
+
+// Check the functional correctness of the summary of safeTransfer.
+rule checkTransferSummary(address token, address to, uint256 amount) {
+    mathint initialBalance = balanceOf(token, currentContract);
+    // Safe require because the total supply is greater than the sum of the balance of any two accounts.
+    require to != currentContract => initialBalance + balanceOf(token, to) <= to_mathint(totalSupply(token));
+
+    libSafeTransfer(token, to, amount);
+    mathint finalBalance = balanceOf(token, currentContract);
+
+    require myBalances[token] == initialBalance;
+    summarySafeTransferFrom(token, currentContract, to, amount);
+    assert myBalances[token] == finalBalance;
+}
+
+// Check the functional correctness of the summary of safeTransferFrom.
+rule checkTransferFromSummary(address token, address from, uint256 amount) {
+    mathint initialBalance = balanceOf(token, currentContract);
+    // Safe require because the total supply is greater than the sum of the balance of any two accounts.
+    require from != currentContract => initialBalance + balanceOf(token, from) <= to_mathint(totalSupply(token));
+
+    libSafeTransferFrom(token, from, currentContract, amount);
+    mathint finalBalance = balanceOf(token, currentContract);
+
+    require myBalances[token] == initialBalance;
+    summarySafeTransferFrom(token, from, currentContract, amount);
+    assert myBalances[token] == finalBalance;
+}
+
+// Check the revert condition of the summary of safeTransfer.
+rule transferRevertCondition(address token, address to, uint256 amount) {
+    uint256 initialBalance = balanceOf(token, currentContract);
+    uint256 toInitialBalance = balanceOf(token, to);
+    // Safe require because the total supply is greater than the sum of the balance of any two accounts.
+    require to != currentContract => initialBalance + toInitialBalance <= to_mathint(totalSupply(token));
+    // Some tokens revert when either the sender or the receiver is the zero address.
+    require currentContract != 0 && to != 0;
```

Comment:
Right, nice catch ! It was the case before when we were using OZ lib for one of the implementation. This dependency was removed because it was a bit overkill just for a dispatch